booster/library/Booster/Pattern/Rewrite.hs

```diff
@@ -299,15 +306,22 @@ applyRule pat@Pattern{ceilConditions} rule = withRuleContext rule $ runRewriteRu
         MatchIndeterminate remainder -> do
             withContext "abort" $
                 logMessage $
-                    renderOneLineText $
-                        "Uncertain about match with rule. Remainder:" <+> pretty remainder
+                    WithJsonMessage (object ["remainder" .= (bimap externaliseTerm externaliseTerm <$> remainder)]) $
+                        renderOneLineText $
+                            "Uncertain about match with rule. Remainder:" <+> pretty remainder
             failRewrite $ RuleApplicationUnclear rule pat.term remainder
         MatchSuccess substitution -> do
-            withContext "success" $
-                logMessage $
-                    renderOneLineText $
-                        "Substitution:"
-                            <+> (hsep $ intersperse "," $ map (\(k, v) -> pretty k <+> "->" <+> pretty v) $ Map.toList substitution)
+            withContext "success" $ do
+                logMessage rule
+                withContext "substitution"
+                    $ logMessage
+                    $ WithJsonMessage
```

Comment:
Do I understand correctly, that `WithJsonMessage` here does not mean that JSON is the only possible output format of this message, i.e. we could still get the normal contextual logs?

Reply:
yes, this just means that the provided json is emitted when json logging is enabled